### PR TITLE
Pythonic job creation

### PR DESCRIPTION
**Pull request by yhwen**
Fixes # .

### Description

Add the support for the pythonic job creartion.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Review by holgerroth — APPROVED** [review]
Nice job. Looks good to me.

File: nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 from typing import Any, Dict, Union
 
 from nvflare.apis.dxo import DXO, DataKind, from_shareable
+from nvflare.apis.event_type import EventType
```
Comment (yanchengnv) [inline nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py:18]: Why do we need to modify this?  It doesn't seem to be related to this PR.

File: nvflare/client/api.py [hunk 2]
```diff
@@ -70,8 +70,8 @@ def send(model: FLModel, clear_cache: bool = True) -> None:
         model (FLModel): Sends a FLModel object.
         clear_cache: clear cache after send
     """
-    if not isinstance(fl_model, FLModel):
-        raise TypeError("fl_model needs to be an instance of FLModel")
+    if not isinstance(model, FLModel):
```
Comment (yanchengnv) [inline nvflare/client/api.py:73]: Why do we need this change?

File: nvflare/job_config/base_app_config.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+# limitations under the License.
+import os.path
+from abc import ABC
+from typing import Dict, List
+
+from nvflare.apis.filter import Filter
+from nvflare.apis.fl_component import FLComponent
+
+
+class BaseAppConfig(ABC):
```
Comment (yanchengnv) [inline nvflare/job_config/base_app_config.py:22]: Please add docstring

File: nvflare/job_config/fed_app_config.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+from nvflare.apis.executor import Executor
+from nvflare.apis.impl.controller import Controller
+from nvflare.apis.impl.wf_comm_server import WFCommServer
+from nvflare.job_config.base_app_config import BaseAppConfig
+from nvflare.private.fed.client.client_json_config import _ExecutorDef
+from nvflare.private.fed.server.server_json_config import WorkFlow
+
+
+class ClientAppConfig(BaseAppConfig):
```
Comment (yanchengnv) [inline nvflare/job_config/fed_app_config.py:24]: Please add docstrings

File: nvflare/job_config/fed_job_config.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+from nvflare.private.fed.app.fl_conf import FL_PACKAGES
+
+CONFIG = "config"
+CUSTOM = "custom"
+FED_SERVER_JSON = "config_fed_server.json"
+FED_CLIENT_JSON = "config_fed_client.json"
+META_JSON = "meta.json"
+
+
+class FedJobConfig:
```
Comment (yanchengnv) [inline nvflare/job_config/fed_job_config.py:32]: Please add docstrings.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Looks good in general. Please see my comments for improvement.

File: nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py (hunk 1 above)
Comment (yhwen) [reply nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py:18]: The old implementation changed the aggregation_weights, exclude_vars, expected_data_kind parameters in the object constructor. When using the inspect to parse the object attributes won't be able to figure out what parameters have been used to create the object, because the parameters have been changed to a different format.

Also performing the data conversion during the START_RUN event is the proper way of implementation.

File: nvflare/client/api.py (hunk 2 above)
Comment (yhwen) [reply nvflare/client/api.py:73]: this is to fix a variable rename error.

File: nvflare/job_config/base_app_config.py (hunk 3 above)
Comment (yhwen) [reply nvflare/job_config/base_app_config.py:22]: Added

File: nvflare/job_config/fed_app_config.py (hunk 4 above)
Comment (yhwen) [reply nvflare/job_config/fed_app_config.py:24]: Added

File: nvflare/job_config/fed_job_config.py (hunk 5 above)
Comment (yhwen) [reply nvflare/job_config/fed_job_config.py:32]: Added

**Review by holgerroth — APPROVED** [review]

File: nvflare/app_opt/pt/in_process_client_api_executor.py [hunk 6]
```diff
@@ -48,7 +49,7 @@ def __init__(
             submit_model_task_name=submit_model_task_name,
             from_nvflare_converter_id=from_nvflare_converter_id,
             to_nvflare_converter_id=to_nvflare_converter_id,
-            params_exchange_format=ExchangeFormat.PYTORCH,
+            params_exchange_format=params_exchange_format,
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:52]: Holger asked to add this. @holgerroth

File: nvflare/app_opt/pt/in_process_client_api_executor.py [hunk 7]
```diff
@@ -36,6 +36,7 @@ def __init__(
         train_task_name: str = "train",
         evaluate_task_name: str = "evaluate",
         submit_model_task_name: str = "submit_model",
+        params_exchange_format=ExchangeFormat.PYTORCH,
```
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:39]: Holger asked to add this. @holgerroth

File: nvflare/job_config/fed_job_config.py (hunk 5 above)
Comment (yhwen) [reply nvflare/job_config/fed_job_config.py:33]: Added docstring the FedJobConfig class.

File: nvflare/app_opt/pt/in_process_client_api_executor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (holgerroth) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:39]: The reason for the change is technical. See my comment below. Also, this is `PTInProcessClientAPIExecutor`, so should it default to `params_exchange_format=ExchangeFormat.PYTORCH`?

File: nvflare/app_opt/pt/in_process_client_api_executor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yhwen) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:52]: In the base class InProcessClientAPIExecutor, it is set to ExchangeFormat.NUMPY. In the sub-class PTInProcessClientAPIExecutor, it is set to use ExchangeFormat.PYTORCH.

[the remark this answers is not in the document]
Comment (holgerroth) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:52]: Why? `PTInProcessClientAPIExecutor` is pytorch-specific, isn't it? L52 is just passing the value on to the super class init call.

[the remark this answers is not in the document]
Comment (holgerroth) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:52]: cc @YuanTingHsieh

File: examples/advanced/job_config/hello-pt/add_shareable_parameter.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (YuanTingHsieh) [inline examples/advanced/job_config/hello-pt/add_shareable_parameter.py:1]: 2024

File: examples/advanced/job_config/hello-pt/cifar10trainer.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (YuanTingHsieh) [inline examples/advanced/job_config/hello-pt/cifar10trainer.py:1]: 2024

File: examples/advanced/job_config/hello-pt/add_shareable_parameter.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+# limitations under the License.
+
+from nvflare.apis.filter import Filter
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+
+
+class AddShareable(Filter):
+    def process(self, shareable: Shareable, fl_ctx: FLContext) -> Shareable:
+        print(f"{fl_ctx.get_identity_name()} ---- AddShareable Filter ----")
```
Comment (YuanTingHsieh) [inline examples/advanced/job_config/hello-pt/add_shareable_parameter.py:22]: This is just printing, should we rename to PrintShareable?

File: examples/advanced/job_config/hello-pt/hello_pt_job.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+        server_app.add_component("json_generator", component)
+
+        task_filter = AddShareable()
+        server_app.add_task_data_filter(["train"], task_filter)
+        task_filter = PrintShareable()
+        server_app.add_task_result_filter(["validate", "train"], task_filter)
+        return server_app
+
+    def export_job(self, job_root):
+        self.job.generate_job_config(job_root)
```
Comment (YuanTingHsieh) [inline examples/advanced/job_config/hello-pt/hello_pt_job.py:110]: Personally prefer 
"Self.job.export"

Like export name better but this is good either way

File: examples/advanced/job_config/hello-pt/simple_network.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
```
Comment (YuanTingHsieh) [inline examples/advanced/job_config/hello-pt/simple_network.py:1]: 2024

File: nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py [hunk 13]
```diff
@@ -15,6 +15,7 @@
 from typing import Any, Dict, Union
```
Comment (YuanTingHsieh) [inline nvflare/app_common/aggregators/intime_accumulate_model_aggregator.py:15]: We might need to move these changes to this file into a separate PR?

File: nvflare/app_common/executors/in_process_client_api_executor.py [hunk 14]
```diff
@@ -69,7 +69,7 @@ def __init__(
             raise ValueError(f"invalid task_script_path '{task_script_path}'")
 
         # only support main() for backward compatibility
-        self._task_fn_path = task_script_path.replace(".py", ".main")
```
Comment (YuanTingHsieh) [inline nvflare/app_common/executors/in_process_client_api_executor.py:72]: Suggest moving in process api enhancement in another PR

File: tests/unit_test/app_common/aggregators/in_time_accumulate_weighted_aggregator_test.py [hunk 15]
```diff
@@ -72,11 +72,14 @@ class TestInTimeAccumulateWeightedAggregator:
     )
     def test_invalid_create(self, exclude_vars, aggregation_weights, expected_data_kind, error, error_msg):
         with pytest.raises(error, match=re.escape(error_msg)):
-            _ = InTimeAccumulateWeightedAggregator(
```
Comment (YuanTingHsieh) [inline tests/unit_test/app_common/aggregators/in_time_accumulate_weighted_aggregator_test.py:75]: Move this into another PR

File: nvflare/job_config/base_app_config.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+        if isinstance(component, FLComponent):
+            self.handlers.append(component)
+
+    def add_task_data_filter(self, tasks: List[str], filter: Filter):
+        self._add_task_filter(tasks, filter, self.task_data_filters)
+
+    def add_task_result_filter(self, tasks: List[str], filter: Filter):
+        self._add_task_filter(tasks, filter, self.task_result_filters)
+
+    def add_ext_script(self, ext_script: str):
```
Comment (YuanTingHsieh) [inline nvflare/job_config/base_app_config.py:53]: Does this belong to BaseApp? Or it is a special ClientApp?

File: nvflare/job_config/fed_job_config.py [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+from typing import Dict
+
+from nvflare import SimulatorRunner
+from nvflare.job_config.fed_app_config import FedAppConfig
+from nvflare.private.fed.app.fl_conf import FL_PACKAGES
+
+CONFIG = "config"
+CUSTOM = "custom"
+FED_SERVER_JSON = "config_fed_server.json"
+FED_CLIENT_JSON = "config_fed_client.json"
```
Comment (YuanTingHsieh) [inline nvflare/job_config/fed_job_config.py:29]: While JSON is OK, we also have .conf file, which one do we prefer @chesterxgchen

File: nvflare/job_config/fed_job_config.py [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+
+    def _generate_meta(self, job_dir):
+        """generate the job meta.json
+
+        Returns:
+
+        """
+        meta_file = os.path.join(job_dir, META_JSON)
+        meta_json = {
+            "name": self.job_name,
```
Comment (YuanTingHsieh) [inline nvflare/job_config/fed_job_config.py:86]: We probably should use some constant variable for these key names

**Review by YuanTingHsieh — CHANGES_REQUESTED** [review]
Mostly good.

I think we should move the enhancement changes into different PRs so it's easier to get them merged

File: nvflare/app_opt/pt/in_process_client_api_executor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:52]: I agree `PTInProcessClientAPIExecutor` is pytorch-specific.

One thing we can think though, we should NOT allow users to pass this `params_exchange_format` for `PTInProcessClientAPIExecutor` to something other than `ExchangeFormat.PYTORCH`?

So if we are making this change, we better add a check in the `__init__` method of `PTInProcessClientAPIExecutor`

File: examples/advanced/job_config/hello-pt/add_shareable_parameter.py (hunk 8 above)
Comment (yhwen) [reply examples/advanced/job_config/hello-pt/add_shareable_parameter.py:1]: changed

File: examples/advanced/job_config/hello-pt/cifar10trainer.py (hunk 9 above)
Comment (yhwen) [reply examples/advanced/job_config/hello-pt/cifar10trainer.py:1]: changed

File: examples/advanced/job_config/hello-pt/add_shareable_parameter.py (hunk 10 above)
Comment (yhwen) [reply examples/advanced/job_config/hello-pt/add_shareable_parameter.py:22]: There's already another PrintShareable in the example. Keep this name.

File: examples/advanced/job_config/hello-pt/hello_pt_job.py (hunk 11 above)
Comment (yhwen) [reply examples/advanced/job_config/hello-pt/hello_pt_job.py:110]: keep for now. We can change later.

File: examples/advanced/job_config/hello-pt/simple_network.py (hunk 12 above)
Comment (yhwen) [reply examples/advanced/job_config/hello-pt/simple_network.py:1]: changed.

File: nvflare/app_common/executors/in_process_client_api_executor.py (hunk 14 above)
Comment (yhwen) [reply nvflare/app_common/executors/in_process_client_api_executor.py:72]: Holger asked to include this, so the example will work with the same PR.

File: tests/unit_test/app_common/aggregators/in_time_accumulate_weighted_aggregator_test.py (hunk 15 above)
Comment (yhwen) [reply tests/unit_test/app_common/aggregators/in_time_accumulate_weighted_aggregator_test.py:75]: Need to have this for unit tests to pass.

File: nvflare/job_config/base_app_config.py (hunk 16 above)
Comment (yhwen) [reply nvflare/job_config/base_app_config.py:53]: This is for the  BaseApp. ServerApp could have it as well.

File: nvflare/job_config/fed_job_config.py (hunk 17 above)
Comment (yhwen) [reply nvflare/job_config/fed_job_config.py:29]: We could add other format of config file support as options as enhancement in the future.

File: nvflare/job_config/fed_job_config.py (hunk 18 above)
Comment (yhwen) [reply nvflare/job_config/fed_job_config.py:86]: We can combine this constant with the job scheduler meta.json parsing in the future.

File: nvflare/app_opt/pt/in_process_client_api_executor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (holgerroth) [reply nvflare/app_opt/pt/in_process_client_api_executor.py:52]: @yhwen , I think it would work if we just remove that argument from PTInProcessClientAPIExecutor, right?

**Review by YuanTingHsieh — APPROVED** [review]
@yhwen please check if you want to update the license header from 2023 to 2024? otherwise you can merge it.

File: examples/advanced/job_config/hello-pt/add_shareable_parameter.py (hunk 8 above)
Comment (YuanTingHsieh) [reply examples/advanced/job_config/hello-pt/add_shareable_parameter.py:1]: has not changed though, but can be in next PR as well

File: examples/advanced/job_config/hello-pt/cifar10trainer.py (hunk 9 above)
Comment (YuanTingHsieh) [reply examples/advanced/job_config/hello-pt/cifar10trainer.py:1]: has not changed though, but can be in next PR as well

File: examples/advanced/job_config/hello-pt/simple_network.py (hunk 12 above)
Comment (YuanTingHsieh) [reply examples/advanced/job_config/hello-pt/simple_network.py:1]: has not changed though, but can be in next PR as well

**Review by YuanTingHsieh — APPROVED** [review]

**Review by yanchengnv — APPROVED** [review]
LGTM
